_pages/dev/tms.md

```diff
@@ -0,0 +1,243 @@
+
+##### Creating a custom mapper
+
+To create a custom mapper per a TMS solution, you can build on top of the [example above](#Data-layer) (by creating a custom collector), and implement the `map` method like so:
+
+```typescript
+import { Injectable } from "@angular/core";
+import { CxEvent } from "@spartacus/core";
+import {
+  TmsCollector,
+  TmsCollectorConfig,
+  WindowObject,
+} from "@spartacus/tracking/tms/core";
+
+@Injectable({ providedIn: "root" })
+export class MyCollectorService implements TmsCollector {
+  ...
+
+  map?<T extends CxEvent>(event: T): T | object {
+    // TODO: custom mapping logic goes here...
+  }
+}
+```
+
+In the `map` method you can use `instanceof` check to differentiate which event should you re-map. As this logic can get quite complex, you have the ability to inject other services to your `MyCollectorService`.
+
+##### Creating custom events
+
+Creating custom events in Spartacus and re-mapping the data to fit your data structure requirements is straightforward:
+
+```typescript
+import { Injectable } from "@angular/core";
+import { createFrom, EventService } from "@spartacus/core";
+import { NavigationEvent } from "@spartacus/storefront";
+import { Observable } from "rxjs";
+import { map } from "rxjs/operators";
+
+export class CustomNavigationEvent {
+  pageType: string;
+  pageName: string;
+}
+
+@Injectable({
+  providedIn: "root",
+})
+export class CustomNavigationEventBuilder {
+  constructor(protected eventService: EventService) {
+    this.buildNavigationEvent();
+  }
+
+  private buildNavigationEvent(): Observable<CustomNavigationEvent> {
+    return this.eventService.get(NavigationEvent).pipe(
+      map((navigationEvent) =>
+        createFrom(CustomNavigationEvent, {
+          // just an example logic, it's not necessarily the actual page name and type
+          pageName: navigationEvent.context.id,
+          pageType: navigationEvent.context.type,
+        })
+      )
+    );
+  }
+}
+```
+
+In the case above, we are re-mapping Spartacus' `NavigationEvent` to `CustomNavigationEvent`. If you need to pull additional data, please see the "Pulling Additional Data From Facades" chapter in the [Events Service docs](./event-service#Pulling-Additional-Data-From-Facades).
+
+Note that you need to inject your `CustomNavigationEventBuilder` builder somewhere, in order to bootstrap the logic - e.g. to a dummy module that's being imported to e.g. `AppModule`.
+
+If you use this approach, you need to pass your `CustomNavigationEvent` to the `tms.adobeLaunch.events` config array, instead of the default `NavigationEvent`.
+
+##### Overriding the TMS service's `mapEvents()`
+
+In case you just want enrich all Spartacus events with some common data, overriding the `TmsService`'s `mapEvents()` is a good place to do it:
+
+```typescript
```

Comment:
Do we need to explain with a complete example how to override a method of a class? ;-)